decisions/2024-03-18_central_server_code_base.md

```diff
@@ -0,0 +1,111 @@
+# Central server code base
+
+- _Date_: 2024-04-03
+- _Deciders_: Andrei Evguenov, James Brunskill, Mark Prins, Chris Petty
+- _Status_: PROPOSED
+- _Outcome_: Option 1
+
+## Context
+
+Until recently open-mSupply relied on a central server using legacy mSupply to sync data.
+This meant that any new entities would require legacy server code changes to handle them.
+
+We want to establish a transitional pattern where we have an open-mSupply Central Server but won't yet have the ability to completely replace the central server with open-mSupply.
+
+Goals:
+
+- Easy to develop, maintain, and test backwards compatibility
+- Have something that's easy to deploy and works nicely alongside the legacy server
+- New functionality can be added to omSupply central rather then adding it to legacy mSupply and having to re-write it again in omSupply in the future
+
+Although we are committed to the existing tech stack, there are various options to consider when implementing omSupply central server. Some of the challanges we have with existing mSupply/mSupply mobile and omSupply ecosystem:
+- New tables and fields need to be added in quite a few places
+- Hard to keep track of compatibility between different versions of central and remote sites of various flavours
+- Challenges in navigating and understand existing mSupply code base due to conditional forks in various places that differentiate functionality based on 'central' or 'remote' or 'single user' variants
+
+## Options
+
+### Option 1 - Single Code Base
+
+By config it's either by a toggle set at start up or by configuration that is set on legacy central server for the site
+
+_Pros:_
+
+- Already implemented (can get going with this straight away)
+- No need to maintain two code bases
+- Would allow a fully standalone server/client setup
+- Able to use rust to check types match between server and client
+
+_Cons:_
+
+- Could be more spaghetti code as you need to have if statements or barriers that check if running as central server
+- Binary & assets might be larger than necessary as both central and remote functions need to be included
+
+#### 1a - Differentiated by config
+
+By config it's either by a toggle set at start up or by configuration that is set on legacy central server for the site
+
+_Pros:_
+
+- Easy to work with in dev (no re-compilation)
+
+_Cons:_
+
+- More configurations during install
+
+#### 1b - Differentiated by feature flag
```

Comment:
Could you clarify how this is different from Option 2? Is this a feature flag that compiles the same project with feature conditions throughout, while Option 2 is a different project importing shared libraries that is built completely separately? (I'm not sure what the right [rust jargon](https://google.github.io/comprehensive-rust/glossary.html) is here.)

Reply:
54e359b18083c2a5898dee32e18a427f6a52eeef
 
Yep basically how you explained it. Option 2 -> composing two different apps from shared, logic is seperate. Option 1b -> Logic is branched in place, can build one or the other based on feature flag/config